crates/kobold_router/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+        };
+
+        // todo!();
+        // let JsValue = view((1, "foo"));
+
+        // self.router
+        //     .borrow_mut()
+        //     .insert(route, JsValue)
+        //     .expect_throw("failed to insert route");
+    }
+
+    /// Starts and hosts your web app with a router
+    pub fn start(&mut self) {
+        kobold::start(view! {
+           <div id="routerView"></div>
+        });
+
+        let local_router = Rc::clone(&self.router);
+
+        let window = web_sys::window().expect_throw("no window");
+        //This is what decides what is render and triggered by pushState
+        let conditonal_router_render: Closure<dyn FnMut()> = Closure::new(
+            move || match local_router.borrow().at(get_path().as_str()) {
+                Ok(Match { value, params }) => {
+                    let params = params
+                        .iter()
+                        .map(|(k, v)| (k.to_owned(), v.to_owned()))
+                        .collect::<Params>();
+                    log_1(&JsValue::from_str(&format!("{:?}", params)));
+                    internal::change_route_view(value);
+                }
+                Err(_) => kobold::start(view! {
+                    <h1> "404" </h1>
+                }),
+            },
+        );
+
+        //Sets up a listener for pushState events
+        internal::setup_push_state_event();
+        window
+            .add_event_listener_with_callback(
+                "pushState",
+                conditonal_router_render.as_ref().unchecked_ref(),
+            )
+            .unwrap();
+        conditonal_router_render.forget();
+
+        navigate(get_path().as_str())
+    }
+}
+
+/// Get the JsValue of a view
+fn get_js_value(view: impl View) -> JsValue {
+    use std::mem::MaybeUninit;
+    use std::pin::pin;
+
+    let product = pin!(MaybeUninit::uninit());
+    let product = In::pinned(product, move |p| view.build(p));
+    //TODO remove need for clone
+    product.js().clone()
+}
+
+/// Navigate to a new path/route
+pub fn navigate(path: &str) {
+    web_sys::window()
+        .expect_throw("no window")
+        .history()
+        .expect_throw("no history")
+        .push_state_with_url(&JsValue::NULL, "", Some(path))
+        .expect_throw("failed to push state");
+}
+
+#[component(class?: "".to_string())]
+pub fn router_link(route: VString, text: VString, class: String) -> impl View + 'static {
+    // A router link component for using kobold_router
+    stateful(
+        || RouteLink::new(route, text, class.into()),
+        |route_link: &Hook<RouteLink>| {
+            let onclick = move |event: MouseEvent<HtmlLinkElement>| {
+                navigate(route_link.route.as_str());
+                event.prevent_default();
+            };
+
+            view! {
+
+                <a class={route_link.class_names.clone()} href={route_link.route.clone()} style="text-decoration: underline;" onclick={onclick}> {&route_link.text}</a>
+            }
+        },
+    )
+}
```

Comment:
Gonna start here :D

Something simple like this would work:

```rust
#[component(class?: "")]
// It's fine to call this just `link` given it's likely to be used a lot I think
pub fn link<'a>(route: &'a str, class: &'a str, children: impl View + 'a) -> impl View + 'a {
    // Only need to turn the string into an owned string here...
    let route = String::from(route);
    
    // ...so we can move it inside the closure giving `onclick`
    // a `'static` lifetime
    let onclick = move |event: MouseEvent<_>| {
        // If you have a function that takes `&str` and you have a `String`
        // or something similar, the idiomatic thing to do is to just pass
        // a simple `&String` reference, and let the compiler deref it.
        navigate(&route);
        event.prevent_default();
    };
    
    view! {
        <a {class} {onclick}>{children}</a>
    }
}
```

Ideally we should avoid doing allocations inside components (the `String::from`, `.into()` when the type is known, etc.) since those are called on every render. I have a  [`fence`](https://docs.rs/kobold/latest/kobold/diff/fn.fence.html) for this, but it currently only works for `View`s, not event listeners, though that should be an easy enough addition I can help you with ("just" need to `IntoListener<E>` for `Fence` but there is a lot of generic types at play there). Then this function could look like this:

```rust
#[component(class?: "")]
pub fn link<'a>(route: &'a str, class: &'a str, children: impl View + 'a) -> impl View + 'a {
    let onclick = fence(route, || {
        let route = String::from(route);

        move |event: MouseEvent<_>| {
            navigate(&route);
            event.prevent_default();
        }
    };
    
    view! {
        <a {class} {onclick}>{children}</a>
    }
}
```

Sidenote: yes that `'a` all over the signature is ugly as hell, but come Rust 2024 edition (should come october) [we'll be able to remove it](https://github.com/rust-lang/rfcs/blob/master/text/3498-lifetime-capture-rules-2024.md)!

Reply:
I appreciate all the details, and it was a big help! I could not get `IntoListener<E>` for `Fence`, so I took the first example you provided. However, I did add a clone to pass the route to the href attribute. I needed that attribute set for the browser's built-in styling. Was not sure if there was a better way or not.